File: packages/runtime-web-three/src/gizmoGeometry.ts
import * as THREE from "three";

export type Vec3Tuple = readonly [number, number, number];
export type RgbTuple = readonly [number, number, number];
export type GizmoKind = "axis" | "wireBox" | "wireSphere";
export type EditorGizmoKind = "bounds" | "camera" | "light" | "transform" | "uiNode";

export interface IGizmoLine {
  color: RgbTuple;
  from: Vec3Tuple;
  to: Vec3Tuple;
}

export interface IGizmoGeometry {
  debugOnly: true;
  kind: GizmoKind;
  lines: IGizmoLine[];
}

export interface IEditorGizmoOverlay {
  debugOnly: true;
  gizmos: Array<IGizmoGeometry & { id: string; role: EditorGizmoKind }>;
}

export function createAxisGizmo(length = 1): IGizmoGeometry {
  const size = Math.max(0, length);
  return {
    debugOnly: true,
    kind: "axis",
    lines: [
      { color: [1, 0, 0], from: [0, 0, 0], to: [size, 0, 0] },
      { color: [0, 1, 0], from: [0, 0, 0], to: [0, size, 0] },
      { color: [0, 0.35, 1], from: [0, 0, 0], to: [0, 0, size] },
    ],
  };
}

export function createWireBoxGizmo(size: Vec3Tuple = [1, 1, 1], color: RgbTuple = [1, 1, 0]): IGizmoGeometry {
  const hx = Math.max(0, size[0]) / 2;
  const hy = Math.max(0, size[1]) / 2;
  const hz = Math.max(0, size[2]) / 2;
  const corners: Vec3Tuple[] = [
    [-hx, -hy, -hz],
    [hx, -hy, -hz],
    [hx, hy, -hz],
    [-hx, hy, -hz],
    [-hx, -hy, hz],
    [hx, -hy, hz],
    [hx, hy, hz],
    [-hx, hy, hz],
  ];
  const edges: Array<readonly [number, number]> = [
    [0, 1],
    [1, 2],
    [2, 3],
    [3, 0],
    [4, 5],
    [5, 6],
    [6, 7],
    [7, 4],
    [0, 4],
    [1, 5],
    [2, 6],
    [3, 7],
  ];
  return {
    debugOnly: true,
    kind: "wireBox",
    lines: edges.map(([from, to]) => ({ color, from: corners[from]!, to: corners[to]! })),
  };
}

export function createWireSphereGizmo(radius = 0.5, segments = 16, color: RgbTuple = [0, 0.75, 1]): IGizmoGeometry {
  const clampedRadius = Math.max(0, radius);
  const count = Math.max(3, Math.floor(segments));
  const lines: IGizmoLine[] = [];
  for (const plane of ["xy", "xz", "yz"] as const) {
    for (let index = 0; index < count; index += 1) {
      const from = circlePoint(plane, clampedRadius, (index / count) * Math.PI * 2);
      const to = circlePoint(plane, clampedRadius, ((index + 1) / count) * Math.PI * 2);
      lines.push({ color, from, to });
    }
  }
  return { debugOnly: true, kind: "wireSphere", lines };
}

export function gizmoToBufferGeometry(gizmo: IGizmoGeometry): THREE.BufferGeometry {
  const positions: number[] = [];
  const colors: number[] = [];
  for (const line of gizmo.lines) {
    positions.push(...line.from, ...line.to);
    colors.push(...line.color, ...line.color);
  }
  const geometry = new THREE.BufferGeometry();
  geometry.setAttribute("position", new THREE.Float32BufferAttribute(positions, 3));
  geometry.setAttribute("color", new THREE.Float32BufferAttribute(colors, 3));
  return geometry;
}

export function buildEditorGizmoOverlay(options: {
  bounds?: readonly { id: string; size?: Vec3Tuple }[];
  cameras?: readonly { id: string }[];
  lights?: readonly { id: string; radius?: number }[];
  transforms?: readonly { id: string; length?: number }[];
  uiNodes?: readonly { id: string; size?: Vec3Tuple }[];
}): IEditorGizmoOverlay {
  const gizmos: IEditorGizmoOverlay["gizmos"] = [];
  for (const item of options.transforms ?? []) {
    gizmos.push({ ...createAxisGizmo(item.length ?? 1), id: item.id, role: "transform" });
  }
  for (const item of options.bounds ?? []) {
    gizmos.push({ ...createWireBoxGizmo(item.size ?? [1, 1, 1], [1, 0.85, 0]), id: item.id, role: "bounds" });
  }
  for (const item of options.cameras ?? []) {
    gizmos.push({ ...createWireBoxGizmo([1.2, 0.7, 1], [0.25, 0.65, 1]), id: item.id, role: "camera" });
  }
  for (const item of options.lights ?? []) {
    gizmos.push({ ...createWireSphereGizmo(item.radius ?? 0.5, 12, [1, 0.95, 0.35]), id: item.id, role: "light" });
  }
  for (const item of options.uiNodes ?? []) {
    gizmos.push({ ...createWireBoxGizmo(item.size ?? [1, 1, 0], [0.9, 0.35, 1]), id: item.id, role: "uiNode" });
  }
  return { debugOnly: true, gizmos: gizmos.sort((left, right) => left.id.localeCompare(right.id)) };
}

function circlePoint(plane: "xy" | "xz" | "yz", radius: number, angle: number): Vec3Tuple {
  const x = Math.cos(angle) * radius;
  const y = Math.sin(angle) * radius;
  if (plane === "xy") {
    return [x, y, 0];
  }
  if (plane === "xz") {
    return [x, 0, y];
  }
  return [0, x, y];
}

File: runtime-bevy/crates/threenative_runtime/src/gizmo_geometry.rs
use bevy::{
    prelude::*,
    render::{mesh::PrimitiveTopology, render_asset::RenderAssetUsages},
};

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum GizmoKind {
    Axis,
    WireBox,
    WireSphere,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct GizmoLine {
    pub color: [f32; 3],
    pub from: [f32; 3],
    pub to: [f32; 3],
}

#[derive(Clone, Debug, PartialEq)]
pub struct GizmoGeometry {
    pub debug_only: bool,
    pub kind: GizmoKind,
    pub lines: Vec<GizmoLine>,
}

pub fn create_axis_gizmo(length: f32) -> GizmoGeometry {
    let size = length.max(0.0);
    GizmoGeometry {
        debug_only: true,
        kind: GizmoKind::Axis,
        lines: vec![
            GizmoLine {
                color: [1.0, 0.0, 0.0],
                from: [0.0, 0.0, 0.0],
                to: [size, 0.0, 0.0],
            },
            GizmoLine {
                color: [0.0, 1.0, 0.0],
                from: [0.0, 0.0, 0.0],
                to: [0.0, size, 0.0],
            },
            GizmoLine {
                color: [0.0, 0.35, 1.0],
                from: [0.0, 0.0, 0.0],
                to: [0.0, 0.0, size],
            },
        ],
    }
}

pub fn create_wire_box_gizmo(size: [f32; 3], color: [f32; 3]) -> GizmoGeometry {
    let hx = size[0].max(0.0) / 2.0;
    let hy = size[1].max(0.0) / 2.0;
    let hz = size[2].max(0.0) / 2.0;
    let corners = [
        [-hx, -hy, -hz],
        [hx, -hy, -hz],
        [hx, hy, -hz],
        [-hx, hy, -hz],
        [-hx, -hy, hz],
        [hx, -hy, hz],
        [hx, hy, hz],
        [-hx, hy, hz],
    ];
    let edges = [
        [0, 1],
        [1, 2],
        [2, 3],
        [3, 0],
        [4, 5],
        [5, 6],
        [6, 7],
        [7, 4],
        [0, 4],
        [1, 5],
        [2, 6],
        [3, 7],
    ];
    GizmoGeometry {
        debug_only: true,
        kind: GizmoKind::WireBox,
        lines: edges
            .into_iter()
            .map(|[from, to]| GizmoLine {
                color,
                from: corners[from],
                to: corners[to],
            })
            .collect(),
    }
}

pub fn create_wire_sphere_gizmo(radius: f32, segments: usize, color: [f32; 3]) -> GizmoGeometry {
    let clamped_radius = radius.max(0.0);
    let count = segments.max(3);
    let mut lines = Vec::new();
    for plane in [CirclePlane::Xy, CirclePlane::Xz, CirclePlane::Yz] {
        for index in 0..count {
            let from = circle_point(
                plane,
                clamped_radius,
                index as f32 / count as f32 * std::f32::consts::TAU,
            );
            let to = circle_point(
                plane,
                clamped_radius,
                (index + 1) as f32 / count as f32 * std::f32::consts::TAU,
            );
            lines.push(GizmoLine { color, from, to });
        }
    }
    GizmoGeometry {
        debug_only: true,
        kind: GizmoKind::WireSphere,
        lines,
    }
}

pub fn gizmo_to_mesh(gizmo: &GizmoGeometry) -> Mesh {
    let mut positions = Vec::new();
    let mut colors = Vec::new();
    for line in &gizmo.lines {
        positions.push(line.from);
        positions.push(line.to);
        colors.push([line.color[0], line.color[1], line.color[2], 1.0]);
        colors.push([line.color[0], line.color[1], line.color[2], 1.0]);
    }
    let mut mesh = Mesh::new(PrimitiveTopology::LineList, RenderAssetUsages::default());
    mesh.insert_attribute(Mesh::ATTRIBUTE_POSITION, positions);
    mesh.insert_attribute(Mesh::ATTRIBUTE_COLOR, colors);
    mesh
}

#[derive(Clone, Copy)]
enum CirclePlane {
    Xy,
    Xz,
    Yz,
}

fn circle_point(plane: CirclePlane, radius: f32, angle: f32) -> [f32; 3] {
    let x = angle.cos() * radius;
    let y = angle.sin() * radius;
    match plane {
        CirclePlane::Xy => [x, y, 0.0],
        CirclePlane::Xz => [x, 0.0, y],
        CirclePlane::Yz => [0.0, x, y],
    }
}
